pkg/scheduler/generic_scheduler.go

```diff
@@ -71,37 +63,34 @@ type ScheduleResult struct {
 	FeasibleNodes int
 }
 
-type genericScheduler struct {
-	cache                    internalcache.Cache
-	nodeInfoSnapshot         *internalcache.Snapshot
-	percentageOfNodesToScore int32
-	nextStartNodeIndex       int
-}
-
 // snapshot snapshots scheduler cache and node infos for all fit and priority
 // functions.
-func (g *genericScheduler) snapshot() error {
+func (sched *Scheduler) snapshot() error {
 	// Used for all fit and priority funcs.
-	return g.cache.UpdateSnapshot(g.nodeInfoSnapshot)
+	return sched.SchedulerCache.UpdateSnapshot(sched.nodeInfoSnapshot)
 }
 
 // Schedule tries to schedule the given pod to one of the nodes in the node list.
 // If it succeeds, it will return the name of the node.
 // If it fails, it will return a FitError error with reasons.
-func (g *genericScheduler) Schedule(ctx context.Context, extenders []framework.Extender, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (result ScheduleResult, err error) {
+func (sched *Scheduler) Schedule(ctx context.Context, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (result ScheduleResult, err error) {
+	return sched.Handle(ctx, sched, fwk, state, pod)
```

Comment:
This looks a little strange here. `sched` invokes Handle and passed itself as the input of this function. 
The purpose we added `Handle` here is used to mockScheduler in test?

Reply:
Yes, it's a little strange. And it's only used to mockScheduler in test for now.

 `scheduleOne()` calls `Schedule()`, but `scheduler_test.go/TestSchedulerScheduleOne` only tests `scheduleOne()` using a fake `Schedule()`. 
I tried to mock a set of parameters to call the real `Schedule()`, but I found it's hard to maintain (for example, add some new tests) and it doesn't fit the purpose of `scheduler_test.go/TestSchedulerScheduleOne`.

Do you have any suggestions to get rid of it?

Reply:
This is indeed strange. Make defaultHandle a member of Scheduler, create `NewScheduler` function and sets handle to defaultHandle; , also ideally avoid exporting all members of Scheduler and add a function to Scheduler that overrides `handle` (which would have the signature `func(ctx context.Context, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (ScheduleResult, error)`)

Reply:
@ahg-g Do you mean that?⬇
```
type scheduler struct {
        name   string
        handle func()
}
func (sched *scheduler) defaultHandle() {
        fmt.Println(sched.name) // Doing Scheduling
}
func (sched *scheduler) setHandle(method func()) {
        sched.handle = method
}
func main() {
        sched := scheduler{name: "hello"}
        sched.handle = sched.defaultHandle
        sched.handle()
}
```
Or We can set a global variable `var handle func(ctx context.Context, sched *scheduler, fwk framework.Framework, state *framework.CycleState, pod *v1.Pod) (ScheduleResult, error))` in scheduler.go, and set this handle to different implement as need, since there will never be two different `scheduler` objects at the same time.

Reply:
Specifically:

```
type scheduler struct {
        name   string
        handle func()
}

func (sched *scheduler) defaultHandle() {
        fmt.Println(sched.name) // Doing Scheduling
}
func (sched *scheduler) setHandle(method func()) {
        sched.handle = method
}

func NewScheduler *scheduler {
  sched := scheduler {
    ....
  }
  sched.handle = sched.defaultHandle
  return &sched
}

// production code
func main {
  sched := NewScheduler{...}
  sched.Schedule()
}

// test code
func Test {
  sched := NewScheduler{...}
  sched.setHandle(fakeHandle)
  sched.Schedule()
}

```

Reply:
Really appreciate for so detail guidance!

I have pushed something new using `newScheduler()`.